core/infosync/infosync.go

```diff
@@ -18,55 +18,135 @@ package infosync
 
 import (
 	"context"
-	"strings"
+	"fmt"
+	"sync"
 
+	"github.com/libp2p/go-libp2p/core/protocol"
+
+	"github.com/obolnetwork/charon/app/featureset"
 	"github.com/obolnetwork/charon/app/log"
 	"github.com/obolnetwork/charon/app/z"
 	"github.com/obolnetwork/charon/core"
 	"github.com/obolnetwork/charon/core/priority"
 )
 
-const topicVersion = "version"
+const (
+	topicVersion  = "version"
+	topicProtocol = "protocol"
+)
 
 // New returns a new infosync component.
-func New(prioritiser *priority.Component, versions []string) *Component {
+func New(prioritiser *priority.Component, versions []string, protocols []protocol.ID) *Component {
+	// Add a mock alpha protocol if alpha features enabled iot to test infosync in prod.
+	// TODO(corver): Remove this once we have an actual use case.
+	if featureset.Enabled(featureset.MockAlpha) {
+		protocols = append(protocols, "/charon/mock_alpha/1.0.0")
+	}
+
+	c := &Component{
+		prioritiser: prioritiser,
+		versions:    versions,
+		protocols:   protocols,
+	}
+
 	prioritiser.Subscribe(func(ctx context.Context, duty core.Duty, results []priority.TopicResult) error {
-		resultType := resultEmpty
+		res := result{slot: duty.Slot}
+		var fields []z.Field
 		for _, result := range results {
-			log.Debug(ctx, "Infosync completed", z.Any(result.Topic, result.Priorities))
-
-			if result.Topic != topicVersion {
-				continue
-			}
+			fields = append(fields, z.Any(result.Topic, result.Priorities))
 
-			if len(result.Priorities) == 0 {
-				resultType = resultEmpty
-			} else if strings.Join(result.PrioritiesOnly(), ",") == strings.Join(versions, ",") {
-				resultType = resultOwn
-			} else {
-				resultType = resultOther
+			for _, prio := range result.PrioritiesOnly() {
+				switch result.Topic {
+				case topicVersion:
+					res.versions = append(res.versions, prio)
+				case topicProtocol:
+					res.protocols = append(res.protocols, protocol.ID(prio))
+				}
 			}
 		}
 
-		completeCounter.WithLabelValues(resultType).Inc()
+		log.Debug(ctx, "Infosync completed", fields...)
+		c.addResult(res)
 
 		return nil
 	})
 
-	return &Component{
-		prioritiser: prioritiser,
-		versions:    versions,
-	}
+	return c
 }
 
 type Component struct {
 	prioritiser *priority.Component
 	versions    []string
+	protocols   []protocol.ID
+
+	mu      sync.Mutex
+	results []result
```

Comment:
do we need to trim this results?

Reply:
don't think so, unless the config changes constantly which it shouldn't

Reply:
limiting to 100 results